File: Algorithms/Python3.x/917-Reverse_Only_Letters.py
# Runtime: 32 ms, faster than 72.65% of Python3 online submissions for Reverse Only Letters.
# Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Reverse Only Letters.
class Solution:
    def reverseOnlyLetters(self, S: str) -> str:
        left, right = 0, len(S) - 1
        S = list(S)
        while left < right:
            while left < right and not S[left].isalpha():
                left += 1
            while left < right and not S[right].isalpha():
                right -= 1
            S[left], S[right] = S[right], S[left]
            left, right = left + 1, right - 1
        return ''.join(S)

File: Algorithms/C++/917-Reverse_Only_Letters.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Reverse Only Letters.
// Memory Usage: 8.7 MB, less than 9.09% of C++ online submissions for Reverse Only Letters.
class Solution {
public:
    string reverseOnlyLetters(string S) {
        stack<char> stk;
        string result;
        for (char ch : S)
            if (isalpha(ch))
                stk.push(ch);
        for (char ch : S)
            if (isalpha(ch)) {
                result.push_back(stk.top());
                stk.pop();
            } else
                result.push_back(ch);
        return result;
    }
};
